astropy/coordinates/sky_coordinate.py

```diff
@@ -484,6 +487,94 @@ def transform_to(self, frame, merge_attributes=True):
             frame_kwargs.pop(attr)
         return self.__class__(new_coord, **frame_kwargs)
 
+    def apply_space_motion(self, new_obstime=None, dt=None):
+        """
+        Update the position of the source represented by this coordinate object
+        to a new specified time using the velocity and assuming linear motion.
+
+        This is sometimes referred to as an "epoch transformation."
+
+        Parameters
+        ----------
+        new_obstime : `~astropy.time.Time`, optional
+            The time at which to evolve the position to.
+        dt : `~astropy.units.Quantity`, `~astropy.time.TimeDelta`, optional
+            An amount of time to evolve the position of the source.
+        """
+
+        if (new_obstime is None and dt is None or
+                new_obstime is not None and dt is not None):
+            raise ValueError("You must specify one of `new_obstime` or `dt`, "
+                             "but not both.")
+
+        # Validate that we have velocity info
+        if 's' not in self.frame.data.differentials:
+            raise ValueError('SkyCoord requires velocity data to evolve the '
+                             'position.')
+
+        if 'obstime' in self.frame.frame_attributes:
+            raise NotImplementedError("Updating the coordinates in a frame "
+                                      "with explicit time dependence is "
+                                      "currently not supported. If you would "
+                                      "like this functionality, please open an "
+                                      "issue on github:\n"
+                                      "https://github.com/astropy/astropy")
+
+        if new_obstime is not None and self.obstime is None:
+            # If no obstime is already on this object, raise an error if a new
+            # obstime is passed: we need to know the time / epoch at which the
+            # the position / velocity were measured initially
+            raise ValueError('This object has no associated `obstime`. '
+                             'apply_space_motion() must receive a time '
+                             'difference, `dt`, and not a new obstime.')
+
+        t1 = self.obstime
+        if t1 is None:
+            # MAGIC NUMBER: if the current SkyCoord object has no obstime,
+            # assume J2000 to do the dt offset.
+            t1 = Time('J2000')
+
+        if dt is None:
+            dt = new_obstime - self.obstime
+        t2 = t1 + dt
+
+        icrs = self.icrs
+        icrs.set_representation_cls(s=SphericalDifferential)
+
+        try:
+            plx = icrs.distance.to(u.arcsecond, u.parallax()).value
```

Comment:
Here and below: the faster way to convert to a value in a different unit is:
```
...to_value(u.arcsecond, u.parallax())
```

Reply:
I don't think time is an argument here, but I guess it makes sense from an API consistency standpoint:
```python
>>> ra = np.random.uniform(100, 200, size=100) * u.deg
>>> dec = np.random.uniform(-30, 30, size=ra.size) * u.deg
>>> distance = np.random.uniform(10, 100, size=ra.size) * u.pc
>>> icrs = coord.ICRS(ra=ra, dec=dec, distance=distance)
>>> %timeit icrs.distance.to(u.arcsecond, u.parallax()).value
169 µs ± 6.95 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
>>> %timeit icrs.distance.to_value(u.arcsecond, u.parallax())
159 µs ± 2.03 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

Reply:
With more points, they might actually be equivalent:
```
206 µs ± 8.62 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
211 µs ± 2.12 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
I actually never use `.to_value()` in my own code so forgot about it - should we be using it in core whenever we need to do this?

Reply:
It mostly matters for fewer points, since it avoids a cast back to Quantity and then on to ndarray.